src/open_inwoner/openzaak/clients.py

```diff
@@ -692,43 +702,207 @@ def fetch_open_tasks(self, bsn: str) -> list[OpenTask]:
-        "zaak": ZakenClient,
-        "catalogi": CatalogiClient,
-        "document": DocumentenClient,
-        "form": FormClient,
+    services_to_client_mapping: Mapping[ZgwClientType, str] = {
+        "zaak": "zrc_service",
+        "catalogi": "ztc_service",
+        "document": "drc_service",
+        "form": "form_service",
     }
-    if client_class := services_to_client_mapping.get(type_):
-        service = getattr(config, f"{type_}_service")
-        if service:
-            client = build_client(
-                service, client_factory=client_class, configured_from=service
-            )
-            return client
 
-    logger.warning("no service defined for %s", type_)
-    return None
+    return [
+        build_zgw_client_from_service(
+            getattr(api_group, services_to_client_mapping[type_])
+        )
+        for api_group in config.api_groups.all()
+    ]
 
 
-def build_zaken_client() -> ZakenClient | None:
-    return _build_zgw_client("zaak")
+_SINGLETON_ZGW_CLIENT_DEPRECATION_MESSAGE = (
+    "Singleton ZGW client factories are in the process of being deprecated in favour of"
+    " multi-ZGW backend aware implementations. Use build_*_clients() or build_zgw_"
+    "client_from_service() instead."
+)
+warnings.filterwarnings(
+    "once", _SINGLETON_ZGW_CLIENT_DEPRECATION_MESSAGE, category=DeprecationWarning
+)
 
 
-def build_catalogi_client() -> CatalogiClient | None:
-    return _build_zgw_client("catalogi")
+def build_zaken_client() -> ZakenClient:
+    warnings.warn(_SINGLETON_ZGW_CLIENT_DEPRECATION_MESSAGE, DeprecationWarning)
+    config = OpenZaakConfig.get_solo()
+    return build_zgw_client_from_service(config.zaak_service)
 
 
-def build_documenten_client() -> DocumentenClient | None:
-    return _build_zgw_client("document")
+def build_zaken_clients() -> list[ZakenClient]:
+    return _build_all_zgw_clients_for_type("zaak")
+
+
+def build_catalogi_client() -> CatalogiClient:
+    warnings.warn(_SINGLETON_ZGW_CLIENT_DEPRECATION_MESSAGE, DeprecationWarning)
+    config = OpenZaakConfig.get_solo()
+    return build_zgw_client_from_service(config.catalogi_service)
+
+
+def build_catalogi_clients() -> list[CatalogiClient]:
+    return _build_all_zgw_clients_for_type("catalogi")
+
+
+def build_documenten_client() -> DocumentenClient:
+    warnings.warn(_SINGLETON_ZGW_CLIENT_DEPRECATION_MESSAGE, DeprecationWarning)
+    config = OpenZaakConfig.get_solo()
+    return build_zgw_client_from_service(config.document_service)
+
+
+def build_documenten_clients() -> list[DocumentenClient]:
+    return _build_all_zgw_clients_for_type("document")
 
 
 def build_forms_client() -> FormClient | None:
-    return _build_zgw_client("form")
+    warnings.warn(_SINGLETON_ZGW_CLIENT_DEPRECATION_MESSAGE, DeprecationWarning)
+    config = OpenZaakConfig.get_solo()
+
+    # Special case: though we require all other services,
+    # the form_service may not in fact be set
+    if not config.form_service:
+        return None
```

Comment:
I wonder if it would be preferable to raise an exception if the attempt is made to build a `form_client` but the `form_service` is not configured.